transmart-core-db/grails-app/services/org/transmartproject/db/tree/TreeService.groovy

```diff
@@ -181,7 +183,30 @@ class TreeService implements TreeResource {
         multiDimensionalDataResource.clearPatientCountCache()
     }
 
-    final private Lock cacheRebuildLock = new ReentrantLock()
+    static class SimpleLock {
+        boolean locked
+    }
+
+    static final private SimpleLock sharedLock = new SimpleLock(locked: false)
+    static final private Lock lockLock = new ReentrantLock()
+
+    boolean tryLock() {
+        lockLock.lock()
```

Comment:
`lockLock`). Is it lock on lock?

Reply:
Indeed. The `lockLock` can not be locked by one thread and unlocked by another. The sharedLock is used for that. The lockLock is to make sure that one thread can check or change the shared lock at the same time.
The scenario is here that the controller thread checks if a rebuild process is currently active. If not, it starts one and activates the lock at the same time.
This spawns a new task with its own thread. That thread has to unlock the lock at the end of the task.

Reply:
I see that I should add documentation in the code.